internal/resolution/variablesources/olm.go

```diff
@@ -0,0 +1,344 @@
+	predicates := []catalogfilter.Predicate[catalogmetadata.Bundle]{
+		catalogfilter.WithPackageName(packageName),
+	}
+
+	if channelName != "" {
+		predicates = append(predicates, catalogfilter.InChannel(channelName))
+	}
+
+	if versionRange != "" {
+		vr, err := mmsemver.NewConstraint(versionRange)
+		if err != nil {
+			return nil, fmt.Errorf("invalid version range '%s': %w", versionRange, err)
+		}
+		predicates = append(predicates, catalogfilter.InMastermindsSemverRange(vr))
+	}
+
+	resultSet := catalogfilter.Filter(allBundles, catalogfilter.And(predicates...))
+	if len(resultSet) == 0 {
+		if versionRange != "" && channelName != "" {
+			return nil, fmt.Errorf("no package '%s' matching version '%s' found in channel '%s'", packageName, versionRange, channelName)
+		}
+		if versionRange != "" {
+			return nil, fmt.Errorf("no package '%s' matching version '%s' found", packageName, versionRange)
+		}
+		if channelName != "" {
+			return nil, fmt.Errorf("no package '%s' found in channel '%s'", packageName, channelName)
+		}
+		return nil, fmt.Errorf("no package '%s' found", packageName)
+	}
+	sort.SliceStable(resultSet, func(i, j int) bool {
+		return catalogsort.ByVersion(resultSet[i], resultSet[j])
+	})
+
+	return olmvariables.NewRequiredPackageVariable(packageName, resultSet), nil
+}
+
+// InstalledPackageVariables returns variables representing packages
+// already installed in the system.
+// Meaning that each BundleDeployment managed by operator-controller
+// has own variable.
+func InstalledPackageVariables(
+	allBundles []*catalogmetadata.Bundle,
+	bundleDeployments []rukpakv1alpha1.BundleDeployment,
+) ([]*olmvariables.InstalledPackageVariable, error) {
+	var successors successorsFunc = legacySemanticsSuccessors
+	if features.OperatorControllerFeatureGate.Enabled(features.ForceSemverUpgradeConstraints) {
+		successors = semverSuccessors
+	}
+
+	result := make([]*olmvariables.InstalledPackageVariable, 0, len(bundleDeployments))
+	processed := sets.Set[string]{}
+	for _, bundleDeployment := range bundleDeployments {
+		sourceImage := bundleDeployment.Spec.Template.Spec.Source.Image
+		if sourceImage == nil || sourceImage.Ref == "" {
+			continue
+		}
+
+		if processed.Has(sourceImage.Ref) {
+			continue
+		}
+		processed.Insert(sourceImage.Ref)
+
+		bundleImage := sourceImage.Ref
+
+		// find corresponding bundle for the installed content
+		resultSet := catalogfilter.Filter(allBundles, catalogfilter.WithBundleImage(bundleImage))
+		if len(resultSet) == 0 {
+			return nil, fmt.Errorf("bundleImage %q not found", bundleImage)
+		}
+
+		// TODO: fast follow - we should check whether we are already supporting the channel attribute in the operator spec.
+		//       if so, we should take the value from spec of the operator CR in the owner ref of the bundle deployment.
+		//       If that channel is set, we need to update the filter above to filter by channel as well.
```

Comment:
This is an old todo. I'm trying to move code around without significant refactoring.

But we have an issue related to this: https://github.com/operator-framework/operator-controller/issues/466